src/cls/cephfs/cls_cephfs.cc

```diff
@@ -0,0 +1,289 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2015 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+
+#include <string>
+#include <errno.h>
+#include <sstream>
+
+#include "objclass/objclass.h"
+
+CLS_VER(1,0)
+CLS_NAME(cephfs_size_scan)
+
+cls_handle_t h_class;
+cls_method_handle_t h_set_if_greater;
+cls_method_handle_t h_accumulate_inode_metadata;
+
+/**
+ * Set a named xattr to a given integer, if and only if the xattr
+ * is not already set to a greater integer.
+ *
+ * If the xattr is missing, or does not encode an integer, then
+ * it is set to the input integer.
+ *
+ * On success, the output buffer is populated with the resulting
+ * integer contained in the xattr.  On failure, a nonzero value is
+ * returned and the contents of the output buffer are undefined.
+ *
+ * @param in: encoded xattr name, uint64_t
+ * @param out: the resulting value of the named xattr
+ */
+static int set_if_greater(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  assert(in != NULL);
+  assert(out != NULL);
+
+  std::string xattr_name;
+  uint64_t input_val = 0;
+
+  // Decode `in`
+  bufferlist::iterator q = in->begin();
+  try {
+    ::decode(xattr_name, q);
+    ::decode(input_val, q);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  // Load existing xattr value
+  uint64_t existing_val = 0;
+  bufferlist existing_bl;
+  bool set_it = false;
+  int r = cls_cxx_getxattr(hctx, xattr_name.c_str(), &existing_bl);
+  if (r == -ENOENT || existing_bl.length() == 0) {
+    set_it = true;
+  } else if (r == 0) {
+    bufferlist::iterator existing_p = existing_bl.begin();
+    try {
+      ::decode(existing_val, existing_p);
+      if (!existing_p.end()) {
+        // Trailing junk?  Consider it invalid and overwrite
+        set_it = true;
```

Comment:
This makes me a bit nervous. We should probably error out if we hit a scenario that we don't think we could have created.